### PR TITLE
Add IPv6 environments and IPv4 and IPv6 test modes

### PR DESCRIPTION
**Pull request by moritzbuhl**
I find this solution not optimal as we have testmode in each of these files.

### CODE REVIEW

File: env-ot12.sh [hunk 1]
```diff
@@ -1,6 +1,8 @@
 LOCAL_IF=ix0
 LOCAL_ADDR=10.3.0.32
 REMOTE_ADDR=10.3.0.33
+LOCAL_ADDR6=fdd7:e83e:66bc::32
+REMOTE_ADDR6=fdd7:e83e:66bc::33
```
Comment (bluhm) [inline env-ot12.sh]: This should be network fdd7:e83e:66bc:0300:: to reflect IPv4 network.

File: env-ot12.sh [hunk 2]
```diff
@@ -1,6 +1,8 @@
...
 REMOTE_SSH=ot13
 
 export LOCAL_IF LOCAL_ADDR REMOTE_ADDR REMOTE_SSH
```
Comment (bluhm) [inline env-ot12.sh]: Forgot exporting the 6 adresses.

File: step.pl [hunk 3]
```diff
@@ -91,16 +93,31 @@
...
 
-@allmodes{qw(all net tcp udp make fs iperf tcpbench udpbench)} = ();
+@allmodes{qw(all net tcp udp make fs iperf tcpbench udpbench iperftcp
+    iperfudp net4 tcp4 udp4 iperf4 tcpbench4 udpbench4 iperftcp4 iperfudp4
+    net6 tcp6 udp6 iperf6 tcpbench6 udpbench6 iperftcp6 iperfudp6)} = ();
 my %testmode = map {
     die "Unknown test mode: $_" unless exists $allmodes{$_};
     $_ => 1;
 } @ARGV;
 $testmode{all} = 1 unless @ARGV;
```
Comment (bluhm) [inline step.pl]: I don't see a reason to make this expansion in perform.pl and step.pl.

File: perform.pl [hunk 4]
```diff
@@ -97,17 +113,42 @@ sub good {
...
     @sshcmd = ('ssh', '-f', $remote_ssh, 'iperf3', '-s', '-D');
     system(@sshcmd)
 	and die "Start iperf3 server with '@sshcmd' failed: $?";
 }
 
-if ($testmode{tcpbench}) {
+if ($testmode{iperftcp6} || $testmode{iperfudp6}) {
+    my @sshcmd = ('ssh', $remote_ssh, 'pkill', 'iperf3');
+    system(@sshcmd);
+    @sshcmd = ('ssh', '-f', $remote_ssh, 'iperf3', '-s', '-D');
```
Comment (bluhm) [inline perform.pl]: You should force iperf3, tcpbench, udpbench to IPv6 mode.  Otherwise these IPv4 tests run twice.

File: perform.pl [hunk 5]
```diff
@@ -30,26 +30,42 @@
...
     die "Unknown test mode: $_" unless exists $allmodes{$_};
     $_ => 1;
 } @ARGV;
 $testmode{all} = 1 unless @ARGV;
-@testmode{qw(net make fs)} = 1..3 if $testmode{all};
-@testmode{qw(tcp udp)} = 1..2 if $testmode{net};
-@testmode{qw(iperftcp iperfudp)} = 1..2 if $testmode{iperf};
-@testmode{qw(iperftcp tcpbench)} = 1..2 if $testmode{tcp};
-@testmode{qw(iperfudp udpbench)} = 1..2 if $testmode{udp};
+@testmode{qw(net make fs)} = 1..4 if $testmode{all};
```
Comment (bluhm) [inline perform.pl]: This list has only 3 entries, so use 1..3

File: perform.pl [hunk 6]
```diff
@@ -30,26 +30,42 @@
...
+@testmode{qw(net4 net6)} = 1..4 if $testmode{net};
```
Comment (bluhm) [inline perform.pl]: 1..2

File: perform.pl [hunk 7]
```diff
@@ -30,26 +30,42 @@
...
+@testmode{qw(iperf4 iperf6)} = 1..2 if $testmode{iperf};
+@testmode{qw(iperftcp4 iperfudp4)} = 1..2 if $testmode{iperf4};
+@testmode{qw(iperftcp6 iperfudp6)} = 1..2 if $testmode{iperf6};
+@testmode{qw(tcp4 tcp6)} = 1..2 if $testmode{tcp};
+@testmode{qw(iperftcp4 tcpbench4)} = 1..2 if $testmode{tcp4};
+@testmode{qw(iperftcp6 tcpbench6)} = 1..2 if $testmode{tcp6};
+@testmode{qw(udp4 udp6)} = 1..2 if $testmode{udp};
+@testmode{qw(iperfudp4 udpbench4)} = 1..2 if $testmode{udp4};
+@testmode{qw(iperfudp6 udpbench6)} = 1..2 if $testmode{udp6};
+@testmode{qw(tcpbench4 tcpbench6)} = 1 if $testmode{tcpbench};
```
Comment (bluhm) [inline perform.pl]: 1..2

File: perform.pl [hunk 8]
```diff
@@ -30,26 +30,42 @@
...
+@testmode{qw(udpbench4 udpbench6)} = 1 if $testmode{udpbench};
```
Comment (bluhm) [inline perform.pl]: 1..2

File: perform.pl (hunk 4 above)
Comment (moritzbuhl) [reply perform.pl]: I also added -4 for the other case.

Comment (moritzbuhl) [reply perform.pl]: should there be a remote_ssh6?

**Comment by moritzbuhl** [comment]
I also added another block to only set ipv4 remote_addrs if *4 is used

File: perform.pl [hunk 9]
```diff
@@ -90,27 +106,57 @@ sub good {
...
-my $local_addr = $ENV{LOCAL_ADDR}
-    or die "Environemnt LOCAL_ADDR not set";
-my $remote_addr = $ENV{REMOTE_ADDR}
-    or die "Environemnt REMOTE_ADDR not set";
 my $remote_ssh = $ENV{REMOTE_SSH}
     or die "Environemnt REMOTE_SSH not set";
 
+if ($testmode{iperftcp4} || $testmode{iperfudp4} ||
+	$testmode{tcpbench4} || $testmode{udpbench4}) {
+    my $local_addr = $ENV{LOCAL_ADDR}
```
Comment (bluhm) [inline perform.pl]: The variable is scoped to the block.  How can it be used after the if block?  Does perl -c compile.
Just set the address and address6 environment unconditionally.

Comment (moritzbuhl) [reply perform.pl]: How about
```perl
my $remote_ssh = $ENV{REMOTE_SSH}
    or die "Environemnt REMOTE_SSH not set";
my $local_addr = $ENV{LOCAL_ADDR};
my $remote_addr = $ENV{REMOTE_ADDR};
my $local_addr6 = $ENV{LOCAL_ADDR6};
my $remote_addr6 = $ENV{REMOTE_ADDR6};

die "Environemnt LOCAL_ADDR and REMOTE_ADDR required"
    if ((!defined $local_addr || !defined $remote_addr) &&
    ($testmode{iperftcp4} || $testmode{iperfudp4} ||
    $testmode{tcpbench4} || $testmode{udpbench4})); 
die "Environemnt LOCAL_ADDR6 and REMOTE_ADDR6 required"
    if ((!defined $local_addr || !defined $remote_addr) &&
    ($testmode{iperftcp6} || $testmode{iperfudp6} ||
    $testmode{tcpbench6} || $testmode{udpbench6}));
```

Comment (bluhm) [reply perform.pl]: Do not make it too complicated.  Just die if any variable is not set.  I configured them on all perform test machines.

**Pull request merged**